Updated authMiddleware docs (#507)
clerk/clerk-docs: docs/references/nextjs/auth-middleware.mdx

With the recommended matcher, all routes are protected by Clerk's authentication middleware, with the exception of internal `/_next/` routes and static files. Static files are detected by matching on paths that end in `.+\..+`. If your middleware needs to run on all routes but you don't want Clerk to protect every route, use the [`ignoredRoutes` option](#options).

### Default behaviour

If there is no `afterAuth` present in the middleware, then `authMiddleware` will fall back to its default behaviour. The default bahaviour is:

- If the [.env based settings](/docs/references/nextjs/custom-signup-signin-pages#update-your-environment-variables) for the sign-in and sign-up routes are present, add those to any routes listed in `publicRoutes`.
- If the .env based settings are not present, add `/sign-in` and `/sign-up` to any routes listed in `publicRoutes`.
- Make all routes from `publicRoutes` public.
- Configure all routes from `ignoredRoutes` to be ignored and return no auth related information.
- Make all remaining routes protected.

#### Example

> Assuming that the .env based settings are for sign-in and sign-up are set to /sign-in and /sign-up resepctively, the following `authMiddleware` would make the routes `/`, `/contact`, `/sign-in` and `/sign-up` public, no routes ignored and all remaining routes protected.

Reviewer: **`sign-in`** and **`sign-up`** need to be formatted like URLs or file names

> Assuming that the .env based settings are for sign-in and sign-up are set to /sign-in and /sign-up resepctively, the following `authMiddleware` would make the routes `/`, `/contact`, `/sign-in` and `/sign-up` public, no routes ignored and all remaining routes protected.

Reviewer: Format .env as a file name

    if (!auth.userId && !auth.isPublicRoute) {
      return redirectToSignIn({ returnBackUrl: req.url });
    }

>     // redirect logged in user to organization selection page if they are not active in an organization

Reviewer: ```suggestion
    // redirect logged in users to organization selection page if they are not active in an organization
```

    if(auth.userId && !auth.orgId && req.nextUrl.pathname !== "/org-selection"){
      const orgSelection = new URL('/org-selection', req.url)
      return NextResponse.redirect(orgSelection)
    }
    // if the user is logged in and trying to access a protected route, allow them to access route
    if (auth.userId && !auth.isPublicRoute) {
      return NextResponse.next()
    }
    // allow users visiting public routes to access them
    return NextResponse.next();
  }
});
```

#### Understanding `afterAuth` configuration

> When using `afterAuth`, the checks must handle all cases—a custom `afterAuth` completely overrides the default behavior covered earlier aside from adding sign-in and sign-up to `publicRoutes`. `afterAuth` does not allow you to add just one check and rely on the default behavior otherwise. 

Reviewer 2: ```suggestion
When using `afterAuth`, the checks must handle all cases — a custom `afterAuth` completely overrides the default behavior covered earlier, aside from adding sign-in and sign-up to `publicRoutes`. `afterAuth` does not allow you to add just one check and rely on the default behavior otherwise. 
```

> When using `afterAuth` the checks must handle all cases -- a custom `afterAuth` completely overrides the default behaviour covered earlier aside from adding sign-in and sign-up to `publicRoutes`. `afterAuth` does not allow you to add just one check and rely on the default behaviour otherwise. 

Reviewer: ```suggestion
When using `afterAuth`, the checks must handle all cases—a custom `afterAuth` completely overrides the default behavior covered earlier aside from adding sign-in and sign-up to `publicRoutes`. `afterAuth` does not allow you to add just one check and rely on the default behavior otherwise. 
```

Reviewer: American English, em dash, and punctuation.

    }
    // Redirect logged in users to organization selection page if they are not active in an organization
    if(auth.userId && !auth.orgId && req.nextUrl.pathname !== "/org-selection"){
      const orgSelection = new URL('/org-selection', req.url)
      return NextResponse.redirect(orgSelection)
    }
    // If the user is logged in and trying to access a protected route, allow them to access route
    if (auth.userId && !auth.isPublicRoute) {
      return NextResponse.next()
    }
    // Allow users visiting public routes to access them
    return NextResponse.next();
  }
});
```

#### Understanding `afterAuth` configuration

When using `afterAuth`, the checks must handle all cases — a custom `afterAuth` completely overrides the default behavior covered earlier, aside from adding sign-in and sign-up to `publicRoutes`. `afterAuth` does not allow you to add just one check and rely on the default behavior otherwise. 

Looking at the example immediately above, the following is happening:

1. Redirect all users who are not logged in but who are visiting a protected page to the sign-in route.
```ts 
if (!auth.userId && !auth.isPublicRoute) {

>       return redirectToSignIn({ returnBackUrl: req.url });
>     }

Reviewer 3: ```suggestion
  return redirectToSignIn({ returnBackUrl: req.url });
}
```

```

> 2. If the user is signed in but is not currently active in an organization and is not visiting the route for selecting an org, then redirect them to the organization selector. This will force users to select and be active in an organization to visit other parts of the site.

Reviewer 3: ```suggestion
2. If the user is signed in but is not currently active in an organization and is not visiting the route for selecting an organization, then redirect them to the organization selector. This will force users to select and be active in an organization to visit other parts of the site.
```

```ts 
if(auth.userId && !auth.orgId && req.nextUrl.pathname !== "/org-selection"){

>       const orgSelection = new URL('/org-selection', req.url)
>       return NextResponse.redirect(orgSelection)
>     }

Reviewer 3: ```suggestion
  const orgSelection = new URL('/org-selection', req.url)
  return NextResponse.redirect(orgSelection)
}
```

```
3. If the user is signed in and accessing a protected route, allow them to access it.
```ts 
 if (auth.userId && !auth.isPublicRoute) {

>       return NextResponse.next()
>     }

Reviewer 3: ```suggestion
  return NextResponse.next()
}
```

```

> 4. If none of the above match, then the user is visiting a public route. Allow them to. Don't check if they are logged in -- it doesn't matter

Reviewer 3: ```suggestion
4. If none of the above match, then the user is visiting a public route. Allow them to. Don't check if they are logged in -- it doesn't matter.
```

};
```

> ### Controlling public routes using `publicRoutes`

Reviewer: I liked "Make pages public..." best... Is there a reason to use "control" instead?

Author: I changed it to 'Making routes public...' I can use 'pages' but I think 'routes' is more accurate.

};
```

> You can use the `req` object to match against urls. The following example will make all routes but `/dashboard` public.

Reviewer: ```suggestion
Use the `req` object to match against urls. The following example makes all routes but `/dashboard` public.
```

```ts filename="middleware.ts" {4}
import { authMiddleware } from "@clerk/nextjs";

export default authMiddleware({
  publicRoutes: req => !req.url.includes('/dashboard')
})

export const config = {
  matcher: ['/((?!.+\\.[\\w]+$|_next).*)', '/', '/(api|trpc)(.*)'],
};
```

```ts filename="middleware.ts"
export default authMiddleware({
  publicRoutes: req => !req.url.includes('/dashboard')
})

> ```

Author: See https://github.com/clerk/clerk-docs/pull/507#discussion_r1405442122. I am asking for thoughts on which to use for this and a few other examples.

> ```

Reviewer 2: gotcha! replied to it :~)

> You can also use regex to match routes. The following is an example of a negative assertion that makes only the `/admin` protected.

Author: For this example and the one below, do we want the full middleware or just the second example with the authMiddleware only?

Author: @rachelnabors and @alexisintech -- any preference with regards to this? Should I got with the longer, full (first) code example or the shorter (second) one only showing the `authMiddleware` code.

Reviewer 2: oh I think the longer !! it should be the full code example, and then highlight the particular lines that represent what the copy talked about (as you have done, which is perfect).

![Execution order of `beforeAuth`, `publicRoutes`, and `afterAuth`](/docs/images/nextjs/middleware-flow-chart.png)

> ## Debugging your middleware

Reviewer: ```suggestion
## Debug your middleware
```

![Execution order of `beforeAuth`, `publicRoutes`, and `afterAuth`](/docs/images/nextjs/middleware-flow-chart.png)

## Debug your middleware

> If you are having issues getting your Middleware dialed in, or are trying to narrow down auth-related issues, you can use the debugging feature in `authMiddlware`. Add `debug: true` to `authMiddlware` and you will get a debug log to your development terminal or your server logs. 

Reviewer 3: ```suggestion
If you are having issues getting your Middleware dialed in, or are trying to narrow down auth-related issues, you can use the debugging feature in `authMiddlware`. Add `debug: true` to `authMiddlware` and you will get debug logs in your terminal. 
```

```ts filename="middleware.ts" {4}
import { authMiddleware } from "@clerk/nextjs";
export default authMiddleware({
  publicRoutes: ["/"]
  debug: true
});

export const config = {

>       matcher: ['/((?!.+\\.[\\w]+$|_next).*)', '/', '/(api|trpc)(.*)'],

Reviewer 3: ```suggestion
  matcher: ['/((?!.+\\.[\\w]+$|_next).*)', '/', '/(api|trpc)(.*)'],
```
